llvm/lib/Target/AIE/AIEAlternateDescriptor.h

```diff
@@ -0,0 +1,74 @@
+//== AIEAlternateDescriptor.h - Define Alternate descriptor Class *-C++-*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+// This file declares the AIEngine Alternate instruction descriptor class
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_SUPPORT_AIEALTERNATEDESCRIPTOR_H
+#define LLVM_SUPPORT_AIEALTERNATEDESCRIPTOR_H
+
+#include "AIEBaseSubtarget.h"
+#include "MCTargetDesc/AIEMCFormats.h"
+
+#include <unordered_map>
+
+namespace llvm {
+
+using MIAltDescsMap = std::unordered_map<MachineInstr *, const MCInstrDesc *>;
+
+class AIEAlternateDescriptor {
+  MIAltDescsMap AlternateDescs;
+
+public:
+  AIEAlternateDescriptor() = default;
+  ~AIEAlternateDescriptor() = default;
+
+  // Construct an alternate descriptor with the given alternate descriptors.
+  AIEAlternateDescriptor(const MIAltDescsMap &AltDescs)
+      : AlternateDescs(AltDescs) {}
+
+  // Set the alternate descriptor for the given multi-opcode instruction.
+  void setAlternateDescriptor(MachineInstr *MI, const unsigned AltInstOpcode) {
+    const AIEBaseSubtarget &STI = AIEBaseSubtarget::get(*MI->getMF());
+    const AIEBaseInstrInfo *TII = STI.getInstrInfo();
+
+    AlternateDescs[MI] = &TII->get(AltInstOpcode);
+  }
+
+  // Return the alternate descriptor for the given multi-opcode instruction.
+  std::optional<const MCInstrDesc *>
+  getAlternateDescriptor(MachineInstr *MI) const {
+    if (auto It = AlternateDescs.find(MI); It != AlternateDescs.end())
+      return It->second;
+    return std::nullopt;
+  }
+
+  const MCInstrDesc *getDesc(MachineInstr *MI) {
+    return getAlternateDescriptor(MI).value_or(&MI->getDesc());
+  }
+
+  // Return the alternate opcode for the given multi-opcode instruction.
+  std::optional<unsigned> getAlternateOpcode(MachineInstr *MI) const {
```

Comment:
I would prefer a name like `getSelectedOpcode` or `getAssignedOpcode`, because it's not just any alternate opcode, it's the one the scheduler selected.

Reply:
Same for `getAlternateDescriptor`

Reply:
(Or to be fully explicit, maybe even `getSelectedAlternateXXXX`, up to you 😄)